auth/provider.go

```diff
@@ -56,26 +61,280 @@ func (r *NoAuthScope) Allow(_ string, _ string) (ok bool) {
+// ensureUsers ensures that the hub users exist and have the necessary roles.
+func (r *Keycloak) ensureUsers(realm *Realm) (err error) {
+	users, err := LoadUsers(Settings.Auth.UserPath)
+	if err != nil {
+		return
+	}
+
+	for _, user := range users {
+		if _, ok := realm.Users[user.Name]; !ok {
+			_true := true
+			u := gocloak.User{
+				Username:        &user.Name,
+				Enabled:         &_true,
+				RequiredActions: &[]string{"UPDATE_PASSWORD"},
+			}
+			userid, kErr := r.client.CreateUser(context.Background(), r.token.AccessToken, r.realm, u)
+			if kErr != nil {
+				err = liberr.Wrap(kErr)
+				return
+			}
+			u.ID = &userid
+			err = r.client.SetPassword(context.Background(), r.token.AccessToken, userid, r.realm, user.Password, true)
+			if err != nil {
+				err = liberr.Wrap(err)
+				return
+			}
+			realm.Users[user.Name] = u
+		}
+		var realmRoles []gocloak.Role
+		for _, role := range user.Roles {
+			realmRole, ok := realm.Roles[role]
+			if !ok {
+				continue
+			}
+			realmRoles = append(realmRoles, realmRole)
+		}
+		err = r.client.AddRealmRoleToUser(context.Background(), r.token.AccessToken, r.realm, *realm.Users[user.Name].ID, realmRoles)
+		if err != nil {
+			err = liberr.Wrap(err)
+			return
+		}
+	}
+
+	return
+}
+
+//
+// ensureRoles ensures that hub roles and scopes are present in keycloak by
+// creating them if they are missing and assigning scope mappings.
+func (r *Keycloak) ensureRoles(realm *Realm) (err error) {
+	hubRoles, err := LoadRoles(Settings.Auth.RolePath)
+	if err != nil {
+		return
+	}
+
+	// create missing roles and scopes, and build mapping of scopes to roles
+	scopesToRoles := make(map[string][]gocloak.Role)
+	for _, role := range hubRoles {
+		if _, ok := realm.Roles[role.Name]; !ok {
+			realmRole := gocloak.Role{Name: &role.Name}
+			id, kErr := r.client.CreateRealmRole(context.Background(), r.token.AccessToken, r.realm, realmRole)
+			if kErr != nil {
+				err = liberr.Wrap(kErr)
+				return
+			}
+			realmRole.ID = &id
+			realm.Roles[role.Name] = realmRole
+			log.Info("Created realm role.", "role", role.Name, "realm", r.realm)
+		}
+
+		for _, res := range role.Resources {
+			for _, verb := range res.Verbs {
+				scopeName := fmt.Sprintf("%s:%s", res.Name, verb)
+				if _, ok := realm.Scopes[scopeName]; !ok {
+					scope := gocloak.ClientScope{Name: &scopeName}
+					id, kErr := r.client.CreateClientScope(context.Background(), r.token.AccessToken, r.realm, scope)
+					if kErr != nil {
+						err = liberr.Wrap(kErr)
+						return
+					}
+					scope.ID = &id
+					realm.Scopes[scopeName] = scope
+					log.Info("Created client scope.", "scope", scopeName, "realm", r.realm)
+				}
+				scopesToRoles[*realm.Scopes[scopeName].ID] = append(scopesToRoles[*realm.Scopes[scopeName].ID], realm.Roles[role.Name])
+			}
+		}
+	}
+
+	// create scope mappings and add default scopes to client
+	idOfClient, err := r.idOfClient(r.id)
+	if err != nil {
+		return
+	}
+	for sid, roles := range scopesToRoles {
```

Comment:
Will `CreateClientScopesScopeMappingsRealmRoles()` and `AddDefaultScopeToClient()` fail of one or more of the mappings already exist?

Reply:
No, this does not appear to fail.